tgfloofbot/plugins/administration/commands.py

```diff
@@ -0,0 +1,117 @@
+
+    class UsernoteCommandArgs(pydantic.BaseModel):
+        bad_user: Optional[str] = pydantic.Field(
+            None, description="A raw user ID")
+        warn_message: Optional[str] = pydantic.Field(
+            None, description="The note")
+
+    @loader.command(name="warn", help="warn a user", admin=True)
+    def warn_command(
+        client: TGFloofbotClient,
+        update: Update,
+        context: CallbackContext,
+        args: WarnCommandArgs
+    ) -> None:
+        warn_helper(client, update, context, args, False)
+        return
+
+    @loader.command(name="note", help="adds a moderation note for a user", admin=True)
+    def usernote_command(
+        client: TGFloofbotClient,
+        update: Update,
+        context: CallbackContext,
+        args: UsernoteCommandArgs
+    ) -> None:
+        warn_helper(client, update, context, args, True)
+        return
+
+    def warn_helper(client, update, context, args, is_note):
+        user = update.effective_user
+        chat = update.effective_chat
+        admin_chat = _resolve_admin_group_id(Update)
+
+        if admin_chat.type != "group":  # group only command
+            context.bot.send_message(chat_id=chat, text="This command can only be used for groups.")
+            return
+        
+        if args.bad_user:
+            try:
+                bad_user = client.updater.bot.get_chat_member(chat_id=admin_chat.id, user_id=args.bad_user).user  # type: ignore
+                if args.warn_message:
+                    reason = args.warn_message
+                else:
+                    reason = "No reason provided."
+            except telegram.error.BadRequest:
+                raise exceptions.UserNotFoundException(args.bad_user)
+                return
+        else: 
+            context.bot.send_message(chat_id=chat.id, text="Invalid arguments. Please provide a user ID/mention and a reason.")
+            return
+
+        try:
+            warning_entry = models.Warning(
+                user_id=bad_user.id,
+                date_added=datetime.datetime.now(timezone.utc),
+                warned_by=user.username,
+                warned_by_id=user.id,
+                reason=reason
+            )
+
+            if is_note:
+                warning_entry.is_usernote = True
+
+            client.db.add(warning_entry)
+            client.db.commit()
+
+        except Exception as e:
+            LOG.error(f"Error occurred when attempting to insert a {'warning' if is_note is False else 'usernote'} into the database: " + str(e))
+            context.bot.send_message(chat_id=chat.id, text=f"A fatal error occurred trying to insert the {'warning' if is_note is False else 'usernote'} into the database.")
+            return
+        if not is_note:
+            try: 
+                context.bot.send_message(chat_id=bad_user.id, text=f"You have been warned in *{admin_chat.title}*\. Reason: *{reason}*", parse_mode="MarkdownV2")
```

Comment:
If the `MarkdownV2` mode is being used, you'll have to escape any markdown in text that's not supposed to have markdown in it.

For this reason, I've made the `em` helper function ("escape markdown").

```suggestion
                context.bot.send_message(
                    chat_id=bad_user.id,
                    text=(
                        f"You have been warned in *{em(admin_chat.title)}*\.\n"
                        f"Reason: *{em(reason)}*"
                    ),
                    parse_mode="MarkdownV2",
                )
```

Reply:
Added